Panic when parse toml slice

POC:

```go
func main() {
	type MyConfig struct {
		Slice []struct {
			A int
		} `json:"slice" toml:"slice"`
	}

	var cfg MyConfig
	loader := aconfig.LoaderFor(&cfg, aconfig.Config{
		// feel free to skip some steps :)
		// SkipDefaults: true,
		// SkipFiles:    true,
		// SkipEnv:      true,
		// SkipFlags:    true,
		EnvPrefix:  "APP",
		FlagPrefix: "app",
		Files:      []string{"config.toml"},
		FileDecoders: map[string]aconfig.FileDecoder{
			// from `aconfigyaml` submodule
			// see submodules in repo for more formats
			".toml": aconfigtoml.New(),
		},
	})

	if err := loader.Load(); err != nil {
		panic(err)
	}
	fmt.Printf("%+v\n", cfg)
}
```

config.toml file:

```toml
[[slice]]
A = 5
```

Result:
```
panic: []map[string]interface {} [map[A:5]]
github.com/cristalhq/aconfig@v0.16.8/reflection.go:190 
```

--- a/aconfig.go
+++ b/aconfig.go
@@ -13,6 +13,7 @@ import (
 type Loader struct {
 	config  Config
 	dst     interface{}
+	parser  *structParser
 	fields  []*fieldData
 	fsys    fs.FS
 	flagSet *flag.FlagSet
@@ -21,6 +22,11 @@ type Loader struct {
 
 // Config to configure configuration loader.
 type Config struct {
+	// NewParser set to true enables a new and better struct parser.
+	// Default is false because there might be bugs.
+	// In the future new parser will be enabled by default.
+	NewParser bool
+
 	SkipDefaults bool // SkipDefaults set to true will not load config from 'default' tag.
 	SkipFiles    bool // SkipFiles set to true will not load config from files.
 	SkipEnv      bool // SkipEnv set to true will not load config from environment variables.
@@ -164,24 +170,37 @@ func (l *Loader) init() {
 		l.config.Args = os.Args[1:]
 	}
 
-	l.fields = l.getFields(l.dst)
+	if l.config.NewParser {
+		l.parser = newStructParser(l.config)
+		if err := l.parser.parseStruct(l.dst); err != nil {
+			l.errInit = err
+			return
+		}
+	} else {
+		l.fields = l.getFields(l.dst)
+	}
 
 	l.flagSet = flag.NewFlagSet(l.config.FlagPrefix, flag.ContinueOnError)
 	if !l.config.SkipFlags {
 		names := make(map[string]bool, len(l.fields))
-		for _, field := range l.fields {
-			flagName := l.fullTag(l.config.FlagPrefix, field, "flag")
-			if flagName == "" {
-				continue
+		if l.config.NewParser {
+			l.flagSet = l.parser.flagSet
+		} else {
+			for _, field := range l.fields {
+				flagName := l.fullTag(l.config.FlagPrefix, field, "flag")
+				if flagName == "" {
+					continue
+				}
+				if names[flagName] && !l.config.AllowDuplicates {
+					l.errInit = fmt.Errorf("duplicate flag %q", flagName)
+					return
+				}
+				names[flagName] = true
+				l.flagSet.String(flagName, field.Tag("default"), field.Tag("usage"))
 			}
-			if names[flagName] && !l.config.AllowDuplicates {
-				l.errInit = fmt.Errorf("duplicate flag %q", flagName)
-				return
-			}
-			names[flagName] = true
-			l.flagSet.String(flagName, field.Tag("default"), field.Tag("usage"))
 		}
 	}
+
 	if l.config.FileFlag != "" {
 		// TODO: should be prefixed ?
 		l.flagSet.String(l.config.FileFlag, "", "config file param")
@@ -257,6 +276,12 @@ func (l *Loader) loadSources() error {
 			return fmt.Errorf("load flags: %w", err)
 		}
 	}
+
+	if l.config.NewParser {
+		if err := l.parser.apply(l.dst); err != nil {
+			return fmt.Errorf("apply: %w", err)
+		}
+	}
 	return nil
 }
 
@@ -278,6 +303,10 @@ func (l *Loader) checkRequired() error {
 }
 
 func (l *Loader) loadDefaults() error {
+	if l.config.NewParser {
+		return nil
+	}
+
 	for _, field := range l.fields {
 		defaultValue := field.Tag("default")
 		if err := l.setFieldData(field, defaultValue); err != nil {
@@ -327,6 +356,13 @@ func (l *Loader) loadFile(file string) error {
 	}
 
 	tag := decoder.Format()
+
+	if l.config.NewParser {
+		if err := l.parser.applyLevel(tag, actualFields); err != nil {
+			return fmt.Errorf("apply %s: %w", tag, err)
+		}
+		return nil
+	}
 
 	for _, field := range l.fields {
 		name := l.fullTag("", field, tag)
@@ -380,6 +416,13 @@ func (l *Loader) loadEnvironment() error {
 	actualEnvs := getEnv(l.config.Envs)
 	dupls := make(map[string]struct{})
 
+	if l.config.NewParser {
+		if err := l.parser.applyFlat("env", actualEnvs); err != nil {
+			return fmt.Errorf("apply env: %w", err)
+		}
+		return nil
+	}
+
 	for _, field := range l.fields {
 		envName := l.fullTag(l.config.EnvPrefix, field, "env")
 		if envName == "" {
@@ -410,6 +453,13 @@ func (l *Loader) postEnvCheck(values map[string]interface{}, dupls map[string]st
 func (l *Loader) loadFlags() error {
 	actualFlags := getFlags(l.flagSet)
 	dupls := make(map[string]struct{})
+
+	if l.config.NewParser {
+		if err := l.parser.applyFlat("flag", actualFlags); err != nil {
+			return fmt.Errorf("apply flag: %w", err)
+		}
+		return nil
+	}
 
 	for _, field := range l.fields {
 		flagName := l.fullTag(l.config.FlagPrefix, field, "flag")

--- a/aconfig.go
+++ b/aconfig.go
@@ -12,7 +12,7 @@ import (
 // Loader of user configuration.
 type Loader struct {
 	config  Config
-	dst     interface{}
+	dst     any
 	parser  *structParser
 	fields  []*fieldData
 	fsys    fs.FS
@@ -103,7 +103,7 @@ type Config struct {
 // FileDecoder is used to read config from files. See aconfig submodules.
 type FileDecoder interface {
 	Format() string
-	DecodeFile(filename string) (map[string]interface{}, error)
+	DecodeFile(filename string) (map[string]any, error)
 	// Init(fsys fs.FS)
 }
 
@@ -122,7 +122,7 @@ type Field interface {
 
 // LoaderFor creates a new Loader based on a given configuration structure.
 // Supports only non-nil structures.
-func LoaderFor(dst interface{}, cfg Config) *Loader {
+func LoaderFor(dst any, cfg Config) *Loader {
 	assertStruct(dst)
 
 	l := &Loader{
@@ -435,7 +435,7 @@ func (l *Loader) loadEnvironment() error {
 	return l.postEnvCheck(actualEnvs, dupls)
 }
 
-func (l *Loader) postEnvCheck(values map[string]interface{}, dupls map[string]struct{}) error {
+func (l *Loader) postEnvCheck(values map[string]any, dupls map[string]struct{}) error {
 	if l.config.AllowUnknownEnvs || l.config.EnvPrefix == "" {
 		return nil
 	}
@@ -473,7 +473,7 @@ func (l *Loader) loadFlags() error {
 	return l.postFlagCheck(actualFlags, dupls)
 }
 
-func (l *Loader) postFlagCheck(values map[string]interface{}, dupls map[string]struct{}) error {
+func (l *Loader) postFlagCheck(values map[string]any, dupls map[string]struct{}) error {
 	if l.config.AllowUnknownFlags || l.config.FlagPrefix == "" {
 		return nil
 	}
@@ -489,7 +489,7 @@ func (l *Loader) postFlagCheck(values map[string]interface{}, dupls map[string]s
 }
 
 // TODO(cristaloleg): revisit.
-func (l *Loader) setField(field *fieldData, name string, values map[string]interface{}, dupls map[string]struct{}) error {
+func (l *Loader) setField(field *fieldData, name string, values map[string]any, dupls map[string]struct{}) error {
 	if !l.config.AllowDuplicates {
 		if _, ok := dupls[name]; ok {
 			return fmt.Errorf("field %q is duplicated", name)

--- a/aconfig_test.go
+++ b/aconfig_test.go
@@ -13,6 +13,72 @@ import (
 	"time"
 )
 
+var newParser = os.Getenv("ACONFIG_NEW") == "true"
+
+func TestTrueSkip(t *testing.T) {
+	var cfg TestConfig
+	loader := LoaderFor(&cfg, Config{
+		NewParser:    newParser,
+		SkipDefaults: true,
+		SkipFiles:    true,
+		SkipEnv:      true,
+		SkipFlags:    true,
+	})
+	if err := loader.Load(); err != nil {
+		t.Fatal(err)
+	}
+
+	want := TestConfig{}
+
+	if have := cfg; !reflect.DeepEqual(have, want) {
+		fmt.Printf("have: %+v\n", *have.Int)
+		t.Fatalf("\nhave: %+v\nwant: %+v", have, want)
+	}
+}
+
+func Test_parse(t *testing.T) {
+	var cfg TestConfig2
+
+	loader := LoaderFor(&cfg, Config{
+		NewParser: newParser,
+		SkipEnv:   true,
+		SkipFlags: true,
+	})
+	if err := loader.Load(); err != nil {
+		t.Fatal(err)
+	}
+
+	// fmt.Printf("\nresult: %+v\n", cfg)
+	// fmt.Printf("b: %v c: %+v\n", *cfg.B, cfg.C)
+}
+
+type TestConfig2 struct {
+	A int    `default:"1"`
+	B *int32 `default:"10" json:"boom_boom"`
+	C *int32 `env:"ccc"`
+	D string `default:"str"`
+	E struct {
+		Bar int    `default:"42"`
+		Foo string `default:"foo"`
+	}
+	F  map[string]int `default:"1:20,3:4"`
+	F2 map[int]string `default:"1:2,3:40"`
+	G  map[string]struct {
+		Baz int `default:"1234"`
+	} // `default:"1:1234"`
+	H  []string            `default:"ab,cd,ef"`
+	H2 []int               `default:"1,2,3"`
+	I  map[string][]string `default:"1:a-b,2:c-d,3:e-f"`
+	J  []struct {
+		Quzz int
+	} //`default:"1,2,3,4"`
+	Y X
+	X
+}
+type X struct {
+	Xex string `default:"XEX" env:"XEXEXE" flag:"axaxa"`
+}
+
 type LogLevel int8
 
 func (l *LogLevel) UnmarshalText(text []byte) error {
@@ -32,8 +98,27 @@ func (l *LogLevel) UnmarshalText(text []byte) error {
 }
 
 func TestDefaults(t *testing.T) {
+	// type TestConfig struct {
+	// 	Str      string `default:"str-def"`
+	// 	Bytes    []byte `default:"bytes-def"`
+	// 	Int      *int32 `default:"123"`
+	// 	HTTPPort int    `default:"8080"`
+	// 	Param    int    // no default tag, so default value
+	// 	ParamPtr *int   // no default tag, so default value
+	// 	Sub      SubConfig
+	// 	Anon     struct {
+	// 		IsAnon bool `default:"true"`
+	// 	}
+	// 	StrSlice []string       `default:"1,2,3" usage:"just pass strings"`
+	// 	Slice    []int          `default:"1,2,3" usage:"just pass elements"`
+	// 	Map1     map[string]int `default:"a:1,b:2,c:3"`
+	// 	Map2     map[int]string `default:"1:a,2:b,3:c"`
+	// 	EmbeddedConfig
+	// }
+
 	var cfg TestConfig
 	loader := LoaderFor(&cfg, Config{
+		NewParser: newParser,
 		SkipFiles: true,
 		SkipEnv:   true,
 		SkipFlags: true,
@@ -45,21 +130,15 @@ func TestDefaults(t *testing.T) {
 		Bytes:    []byte("bytes-def"),
 		Int:      int32Ptr(123),
 		HTTPPort: 8080,
-		Sub: SubConfig{
-			Float: 123.123,
-		},
+		Sub:      SubConfig{Float: 123.123},
 		Anon: struct {
 			IsAnon bool `default:"true"`
-		}{
-			IsAnon: true,
-		},
-		StrSlice: []string{"1", "2", "3"},
-		Slice:    []int{1, 2, 3},
-		Map1:     map[string]int{"a": 1, "b": 2, "c": 3},
-		Map2:     map[int]string{1: "a", 2: "b", 3: "c"},
-		EmbeddedConfig: EmbeddedConfig{
-			Em: "em-def",
-		},
+		}{IsAnon: true},
+		StrSlice:       []string{"1", "2", "3"},
+		Slice:          []int{1, 2, 3},
+		Map1:           map[string]int{"a": 1, "b": 2, "c": 3},
+		Map2:           map[int]string{1: "a", 2: "b", 3: "c"},
+		EmbeddedConfig: EmbeddedConfig{Em: "em-def"},
 	}
 	mustEqual(t, cfg, want)
 }
@@ -84,14 +163,15 @@ func TestDefaults_AllTypes(t *testing.T) {
 		Float32 float32 `default:"1234.213"`
 		Float64 float64 `default:"1234.234"`
 
-		Dur  time.Duration `default:"1h2m3s"`
-		Time time.Time     `default:"2000-04-05 10:20:30 +0000 UTC"`
+		Dur time.Duration `default:"1h2m3s"`
+		// Time time.Time     `default:"2000-04-05 10:20:30 +0000 UTC"`
 
 		Level LogLevel `default:"warn"`
 	}
 
 	var cfg AllTypesConfig
 	loader := LoaderFor(&cfg, Config{
+		NewParser: newParser,
 		SkipFiles: true,
 		SkipEnv:   true,
 		SkipFlags: true,
@@ -136,6 +216,7 @@ func TestDefaults_OtherNumberFormats(t *testing.T) {
 
 	var cfg OtherNumberFormats
 	loader := LoaderFor(&cfg, Config{
+		NewParser: newParser,
 		SkipFiles: true,
 		SkipEnv:   true,
 		SkipFlags: true,
@@ -161,6 +242,7 @@ func TestJSON(t *testing.T) {
 
 	var cfg structConfig
 	loader := LoaderFor(&cfg, Config{
+		NewParser:    newParser,
 		SkipDefaults: true,
 		SkipEnv:      true,
 		SkipFlags:    true,
@@ -180,6 +262,7 @@ func TestJSONWithOmitempty(t *testing.T) {
 		APIKey string `json:"b,omitempty"`
 	}
 	loader := LoaderFor(&cfg, Config{
+		NewParser:          newParser,
 		SkipDefaults:       true,
 		SkipEnv:            true,
 		SkipFlags:          true,
@@ -195,6 +278,7 @@ func TestCustomFile(t *testing.T) {
 
 	var cfg structConfig
 	loader := LoaderFor(&cfg, Config{
+		NewParser:    newParser,
 		SkipDefaults: true,
 		SkipEnv:      true,
 		SkipFlags:    true,
@@ -215,6 +299,7 @@ func TestFile(t *testing.T) {
 
 	var cfg TestConfig
 	loader := LoaderFor(&cfg, Config{
+		NewParser:    newParser,
 		SkipDefaults: true,
 		SkipEnv:      true,
 		SkipFlags:    true,
@@ -247,6 +332,7 @@ func TestFileEmbed(t *testing.T) {
 
 	var cfg TestConfig
 	loader := LoaderFor(&cfg, Config{
+		NewParser:    newParser,
 		SkipDefaults: true,
 		SkipEnv:      true,
 		SkipFlags:    true,
@@ -279,6 +365,7 @@ func TestFileMerging(t *testing.T) {
 
 	var cfg TestConfig
 	loader := LoaderFor(&cfg, Config{
+		NewParser:    newParser,
 		SkipDefaults: true,
 		SkipEnv:      true,
 		SkipFlags:    true,
@@ -306,6 +393,7 @@ func TestFileFlag(t *testing.T) {
 
 	var cfg TestConfig
 	loader := LoaderFor(&cfg, Config{
+		NewParser:    newParser,
 		SkipDefaults: true,
 		SkipEnv:      true,
 		MergeFiles:   true,
@@ -329,6 +417,7 @@ func TestBadFileFlag(t *testing.T) {
 
 	var cfg TestConfig
 	loader := LoaderFor(&cfg, Config{
+		NewParser:    newParser,
 		SkipDefaults: true,
 		SkipEnv:      true,
 		FileFlag:     "file_flag",
@@ -342,6 +431,7 @@ func TestNoFileFlagValue(t *testing.T) {
 
 	var cfg TestConfig
 	loader := LoaderFor(&cfg, Config{
+		NewParser:    newParser,
 		SkipDefaults: true,
 		SkipEnv:      true,
 		FileFlag:     "file_flag",
@@ -367,8 +457,12 @@ func TestEnv(t *testing.T) {
 	t.Setenv("TST_EM", "em-env")
 	defer os.Clearenv()
 
+	// type TestConfig struct {
+	// 	Sub SubConfig
+	// }
 	var cfg TestConfig
 	loader := LoaderFor(&cfg, Config{
+		NewParser:    newParser,
 		SkipDefaults: true,
 		SkipFiles:    true,
 		SkipFlags:    true,
@@ -393,13 +487,13 @@ func TestEnv(t *testing.T) {
 			Em: "em-env",
 		},
 	}
-
 	mustEqual(t, cfg, want)
 }
 
 func TestFlag(t *testing.T) {
 	var cfg TestConfig
 	loader := LoaderFor(&cfg, Config{
+		NewParser:    newParser,
 		SkipDefaults: true,
 		SkipFiles:    true,
 		SkipEnv:      true,
@@ -437,7 +531,6 @@ func TestFlag(t *testing.T) {
 			Em: "em-flag",
 		},
 	}
-
 	mustEqual(t, cfg, want)
 }
 
@@ -456,6 +549,7 @@ func TestExactName(t *testing.T) {
 	var cfg ExactConfig
 
 	loader := LoaderFor(&cfg, Config{
+		NewParser:        newParser,
 		SkipDefaults:     true,
 		SkipFiles:        true,
 		SkipFlags:        true,
@@ -470,7 +564,6 @@ func TestExactName(t *testing.T) {
 		},
 		Bar: "bar-env",
 	}
-
 	mustEqual(t, cfg, want)
 }
 
@@ -480,7 +573,7 @@ func TestSkipName(t *testing.T) {
 	defer os.Clearenv()
 
 	type Foo struct {
-		String string `env:"STR"`
+		String string `default:"str" env:"STR"`
 	}
 	type ExactConfig struct {
 		Foo Foo    `env:"-"`
@@ -489,6 +582,7 @@ func TestSkipName(t *testing.T) {
 	var cfg ExactConfig
 
 	loader := LoaderFor(&cfg, Config{
+		NewParser: newParser,
 		SkipFiles: true,
 		SkipFlags: true,
 	})
@@ -500,7 +594,6 @@ func TestSkipName(t *testing.T) {
 		},
 		Bar: "def",
 	}
-
 	mustEqual(t, cfg, want)
 }
 
@@ -518,6 +611,7 @@ func TestDuplicatedName(t *testing.T) {
 	var cfg ExactConfig
 
 	loader := LoaderFor(&cfg, Config{
+		NewParser:       newParser,
 		SkipFlags:       true,
 		AllowDuplicates: true,
 	})
@@ -529,7 +623,6 @@ func TestDuplicatedName(t *testing.T) {
 		},
 		FooBar: "str-env",
 	}
-
 	mustEqual(t, cfg, want)
 }
 
@@ -544,6 +637,7 @@ func TestFailOnDuplicatedName(t *testing.T) {
 	var cfg ExactConfig
 
 	loader := LoaderFor(&cfg, Config{
+		NewParser: newParser,
 		SkipFlags: true,
 	})
 
@@ -561,7 +655,7 @@ func TestFailOnDuplicatedFlag(t *testing.T) {
 		Baz string `flag:"yes"`
 	}
 
-	err := LoaderFor(&Foo{}, Config{}).Load()
+	err := LoaderFor(&Foo{}, Config{NewParser: newParser}).Load()
 	failIfOk(t, err)
 
 	want := `init loader: duplicate flag "yes"`
@@ -569,19 +663,20 @@ func TestFailOnDuplicatedFlag(t *testing.T) {
 }
 
 func TestUsage(t *testing.T) {
-	loader := LoaderFor(&EmbeddedConfig{}, Config{})
+	loader := LoaderFor(&EmbeddedConfig{}, Config{
+		NewParser: newParser,
+	})
 
 	var builder strings.Builder
 	flags := loader.Flags()
 	flags.SetOutput(&builder)
 	flags.PrintDefaults()
 
-	got := builder.String()
+	have := builder.String()
 	want := `  -em string
     	use... em...field. (default "em-def")
 `
-
-	mustEqual(t, got, want)
+	mustEqual(t, have, want)
 }
 
 func TestBadDefauts(t *testing.T) {
@@ -589,6 +684,7 @@ func TestBadDefauts(t *testing.T) {
 		t.Helper()
 
 		loader := LoaderFor(cfg, Config{
+			NewParser: newParser,
 			SkipFiles: true,
 			SkipEnv:   true,
 			SkipFlags: true,
@@ -657,11 +753,11 @@ func TestBadDefauts(t *testing.T) {
 	}{})
 
 	f(&struct {
-		Map map[string]int `default:"1:a,2:2"`
+		Map map[string]int `default:"1:a;2:2"`
 	}{})
 
 	f(&struct {
-		Map map[int]string `default:"a:1"`
+		Map map[int]string `default:"a:1;"`
 	}{})
 
 	f(&struct {
@@ -680,6 +776,7 @@ func TestBadFiles(t *testing.T) {
 			t.Helper()
 			var cfg TestConfig
 			loader := LoaderFor(&cfg, Config{
+				NewParser:          newParser,
 				SkipDefaults:       true,
 				SkipEnv:            true,
 				SkipFlags:          true,
@@ -704,6 +801,7 @@ func TestFailOnFileNotFound(t *testing.T) {
 		t.Helper()
 
 		loader := LoaderFor(&TestConfig{}, Config{
+			NewParser:          newParser,
 			SkipDefaults:       true,
 			SkipEnv:            true,
 			SkipFlags:          true,
@@ -723,6 +821,7 @@ func TestBadEnvs(t *testing.T) {
 	defer os.Clearenv()
 
 	loader := LoaderFor(&TestConfig{}, Config{
+		NewParser:    newParser,
 		SkipDefaults: true,
 		SkipFiles:    true,
 		SkipFlags:    true,
@@ -734,6 +833,7 @@ func TestBadEnvs(t *testing.T) {
 
 func TestBadFlags(t *testing.T) {
 	loader := LoaderFor(&TestConfig{}, Config{
+		NewParser:    newParser,
 		SkipDefaults: true,
 		SkipFiles:    true,
 		SkipEnv:      true,
@@ -751,6 +851,7 @@ func TestUnknownFields(t *testing.T) {
 
 	var cfg TestConfig
 	loader := LoaderFor(&cfg, Config{
+		NewParser:    newParser,
 		SkipDefaults: true,
 		SkipEnv:      true,
 		SkipFlags:    true,
@@ -773,6 +874,7 @@ func TestUnknownEnvs(t *testing.T) {
 
 	var cfg TestConfig
 	loader := LoaderFor(&cfg, Config{
+		NewParser:    newParser,
 		SkipDefaults: true,
 		SkipFiles:    true,
 		SkipFlags:    true,
@@ -794,6 +896,7 @@ func TestUnknownEnvsWithEmptyPrefix(t *testing.T) {
 
 	var cfg TestConfig
 	loader := LoaderFor(&cfg, Config{
+		NewParser:    newParser,
 		SkipDefaults: true,
 		SkipFiles:    true,
 		SkipFlags:    true,
@@ -804,6 +907,7 @@ func TestUnknownEnvsWithEmptyPrefix(t *testing.T) {
 
 func TestUnknownFlags(t *testing.T) {
 	loader := LoaderFor(&TestConfig{}, Config{
+		NewParser:    newParser,
 		SkipDefaults: true,
 		SkipFiles:    true,
 		SkipEnv:      true,
@@ -836,6 +940,7 @@ func TestUnknownFlags(t *testing.T) {
 
 func TestUnknownFlagsWithEmptyPrefix(t *testing.T) {
 	loader := LoaderFor(&TestConfig{}, Config{
+		NewParser:    newParser,
 		SkipDefaults: true,
 		SkipFiles:    true,
 		SkipEnv:      true,
@@ -860,6 +965,7 @@ func TestUnknownFlagsWithEmptyPrefix(t *testing.T) {
 // flag.FlagSet already fails on undefined flag.
 func TestUnknownFlagsStdlib(t *testing.T) {
 	loader := LoaderFor(&TestConfig{}, Config{
+		NewParser:    newParser,
 		SkipDefaults: true,
 		SkipFiles:    true,
 		SkipEnv:      true,
@@ -908,7 +1014,8 @@ func TestCustomNames(t *testing.T) {
 
 	var cfg TestConfig
 	loader := LoaderFor(&cfg, Config{
-		Args: []string{"-two=2", "-four=4"},
+		NewParser: newParser,
+		Args:      []string{"-two=2", "-four=4"},
 	})
 
 	failIfErr(t, loader.Load())
@@ -943,6 +1050,7 @@ func TestDontGenerateTags(t *testing.T) {
 	}
 	cfg := Config{
 		DontGenerateTags: true,
+		NewParser:        newParser,
 	}
 	LoaderFor(&testConfig{}, cfg).WalkFields(func(f Field) bool {
 		for _, tag := range []string{"json", "yaml", "env", "flag"} {
@@ -957,6 +1065,9 @@ func TestDontGenerateTags(t *testing.T) {
 }
 
 func TestWalkFields(t *testing.T) {
+	if newParser {
+		t.Skip()
+	}
 	type TestConfig struct {
 		A int `default:"-1" env:"one" marco:"polo"`
 		B struct {
@@ -997,7 +1108,7 @@ func TestWalkFields(t *testing.T) {
 
 	i := 0
 
-	LoaderFor(&TestConfig{}, Config{}).WalkFields(func(f Field) bool {
+	LoaderFor(&TestConfig{}, Config{NewParser: newParser}).WalkFields(func(f Field) bool {
 		wantFields := fields[i]
 		mustEqual(t, f.Name(), wantFields.Name)
 		mustEqual(t, f.Name(), wantFields.Name)
@@ -1013,7 +1124,7 @@ func TestWalkFields(t *testing.T) {
 	mustEqual(t, i, 3)
 
 	i = 0
-	LoaderFor(&TestConfig{}, Config{}).WalkFields(func(f Field) bool {
+	LoaderFor(&TestConfig{}, Config{NewParser: newParser}).WalkFields(func(f Field) bool {
 		if i > 0 {
 			return false
 		}
@@ -1030,6 +1141,7 @@ func TestWalkFields(t *testing.T) {
 
 func TestDontFillFlagsIfDisabled(t *testing.T) {
 	loader := LoaderFor(&TestConfig{}, Config{
+		NewParser: newParser,
 		SkipFlags: true,
 		Args:      []string{},
 	})
@@ -1051,7 +1163,9 @@ func TestPassBadStructs(t *testing.T) {
 			}
 		}()
 
-		_ = LoaderFor(cfg, Config{})
+		_ = LoaderFor(cfg, Config{
+			NewParser: newParser,
+		})
 	}
 
 	f(nil)
@@ -1081,7 +1195,9 @@ func TestBadRequiredTag(t *testing.T) {
 			}
 		}()
 
-		_ = LoaderFor(cfg, Config{})
+		_ = LoaderFor(cfg, Config{
+			NewParser: newParser,
+		})
 	}
 
 	f(&TestConfig{})
@@ -1130,8 +1246,9 @@ type TestConfig struct {
 	Int      *int32 `default:"123"`
 	HTTPPort int    `default:"8080"`
 	Param    int    // no default tag, so default value
-	Sub      SubConfig
-	Anon     struct {
+	// ParamPtr *int   // no default tag, so default value
+	Sub  SubConfig
+	Anon struct {
 		IsAnon bool `default:"true"`
 	}
 
@@ -1163,7 +1280,7 @@ type structConfig struct {
 	AA structA `json:"A"`
 	StructM
 
-	M interface{} `json:"M"`
+	MM interface{} `json:"MM"`
 
 	P *structP `json:"P"`
 }
@@ -1230,7 +1347,7 @@ var testfile = &fstest.MapFile{Data: []byte(`{
 
 	"m": "n",
 
-	"M":["q", "w"],
+	"MM":["q", "w"],
 	
 	"P": {
 		"P": "r"
@@ -1273,7 +1390,7 @@ var wantConfig = func() structConfig {
 		StructM: StructM{
 			M: "n",
 		},
-		M: mInterface,
+		MM: mInterface,
 		P: &structP{
 			P: "r",
 		},
@@ -1305,6 +1422,7 @@ type ConfigVCenterDC struct {
 func TestSliceStructs(t *testing.T) {
 	var cfg ConfigTest
 	loader := LoaderFor(&cfg, Config{
+		NewParser:    newParser,
 		SkipDefaults: true,
 		SkipEnv:      true,
 		SkipFlags:    true,
@@ -1335,13 +1453,14 @@ func TestSliceStructs(t *testing.T) {
 	mustEqual(t, cfg, want)
 }
 
-func TestMapOfMap(t *testing.T) {
+func TestJSONMap(t *testing.T) {
 	type TestConfig struct {
 		Options map[string]float64
 	}
 	var cfg TestConfig
 
 	loader := LoaderFor(&cfg, Config{
+		NewParser:    newParser,
 		SkipDefaults: true,
 		SkipEnv:      true,
 		SkipFlags:    true,
@@ -1361,25 +1480,29 @@ func TestMapOfMap(t *testing.T) {
 }
 
 func TestBad(t *testing.T) {
+	t.Skip("probably too picky")
+
 	type TestConfig struct {
 		Params url.Values
 	}
 	var cfg TestConfig
 	t.Setenv("PARAMS", "foo:bar")
 
-	loader := LoaderFor(&cfg, Config{
-		SkipFlags: true,
-	})
-	failIfErr(t, loader.Load())
-
 	p, err := url.ParseQuery("foo=bar")
 	if err != nil {
 		t.Fatal(err)
 	}
+	fmt.Printf("have: %+v\n", p)
+
+	loader := LoaderFor(&cfg, Config{
+		NewParser: newParser,
+		SkipFlags: true,
+	})
+	failIfErr(t, loader.Load())
+
 	want := TestConfig{
 		Params: p,
 	}
-
 	mustEqual(t, cfg, want)
 }
 
@@ -1393,12 +1516,12 @@ func TestFileConfigFlagDelim(t *testing.T) {
 	var cfg TestConfig
 
 	loader := LoaderFor(&cfg, Config{
+		NewParser:     newParser,
 		SkipDefaults:  true,
 		SkipEnv:       true,
 		SkipFlags:     true,
 		FlagDelimiter: "_",
-
-		Files: []string{"testdata/toy.json"},
+		Files:         []string{"testdata/toy.json"},
 	})
 
 	failIfErr(t, loader.Load())

--- a/aconfig_test.go
+++ b/aconfig_test.go
@@ -680,7 +680,7 @@ func TestUsage(t *testing.T) {
 }
 
 func TestBadDefauts(t *testing.T) {
-	f := func(cfg interface{}) {
+	f := func(cfg any) {
 		t.Helper()
 
 		loader := LoaderFor(cfg, Config{
@@ -1153,7 +1153,7 @@ func TestDontFillFlagsIfDisabled(t *testing.T) {
 }
 
 func TestPassBadStructs(t *testing.T) {
-	f := func(cfg interface{}) {
+	f := func(cfg any) {
 		t.Helper()
 
 		defer func() {
@@ -1185,7 +1185,7 @@ func TestBadRequiredTag(t *testing.T) {
 		Field string `required:"boom"`
 	}
 
-	f := func(cfg interface{}) {
+	f := func(cfg any) {
 		t.Helper()
 
 		defer func() {
@@ -1280,7 +1280,7 @@ type structConfig struct {
 	AA structA `json:"A"`
 	StructM
 
-	MM interface{} `json:"MM"`
+	MM any `json:"MM"`
 
 	P *structP `json:"P"`
 }
@@ -1358,7 +1358,7 @@ var testfile = &fstest.MapFile{Data: []byte(`{
 var wantConfig = func() structConfig {
 	i := int32(42)
 	j := int64(420)
-	mInterface := make([]interface{}, 2)
+	mInterface := make([]any, 2)
 	for iI, vI := range []string{"q", "w"} {
 		mInterface[iI] = vI
 	}
@@ -1582,7 +1582,7 @@ func failIfErr(tb testing.TB, err error) {
 	}
 }
 
-func mustEqual(tb testing.TB, got, want interface{}) {
+func mustEqual(tb testing.TB, got, want any) {
 	tb.Helper()
 	if !reflect.DeepEqual(got, want) {
 		tb.Fatalf("\nhave %+v\nwant %+v", got, want)

--- a/go.mod
+++ b/go.mod
@@ -1,3 +1,5 @@
 module github.com/cristalhq/aconfig
 
 go 1.16
+
+require github.com/mitchellh/mapstructure v1.5.0

--- a/go.mod
+++ b/go.mod
@@ -1,5 +1,5 @@
 module github.com/cristalhq/aconfig
 
-go 1.16
+go 1.18
 
 require github.com/mitchellh/mapstructure v1.5.0

--- a/parser.go
+++ b/parser.go
@@ -1,0 +1,525 @@
+package aconfig
+
+import (
+	"encoding"
+	"flag"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/mitchellh/mapstructure"
+)
+
+type structParser struct {
+	dst       interface{}
+	cfg       Config
+	fields    map[string]interface{}
+	flagSet   *flag.FlagSet
+	envNames  map[string]struct{}
+	flagNames map[string]struct{}
+}
+
+func newStructParser(cfg Config) *structParser {
+	return &structParser{
+		cfg:       cfg,
+		flagSet:   flag.NewFlagSet(cfg.FlagPrefix, flag.ContinueOnError),
+		envNames:  map[string]struct{}{},
+		flagNames: map[string]struct{}{},
+	}
+}
+
+type parsedField struct {
+	name         string
+	namefull     string
+	value        interface{}
+	defaultValue interface{}
+	parent       *parsedField
+	childs       map[string]interface{}
+	tags         map[string]string
+	hasChilds    bool
+	isRequired   bool
+	isSet        bool
+}
+
+func (pf *parsedField) String() string {
+	if pf == nil {
+		return "<nil-pf>"
+	}
+	return fmt.Sprintf("%+v", *pf)
+}
+
+func (sp *structParser) newParseField(parent *parsedField, field reflect.StructField) (*parsedField, error) {
+	requiredTag := field.Tag.Get("required")
+	if requiredTag != "" && requiredTag != "true" {
+		panic(fmt.Sprintf("aconfig: value for 'required' tag can be only 'true' got: %q", requiredTag))
+	}
+
+	name := field.Tag.Get("name")
+	if name == "" {
+		name = field.Name
+	}
+
+	newName := strings.ToLower(strings.Join(splitNameByWords(name), "_"))
+
+	env := field.Tag.Get("env")
+	if env == "" {
+		env = strings.ToUpper(newName)
+	}
+
+	flag := field.Tag.Get("flag")
+	if flag == "" {
+		flag = newName
+	}
+
+	var parentName, parentEnv, parentFlag string
+	if parent != nil {
+		parentName = parent.namefull + "|"
+
+		for p := parent; p != nil; p = p.parent {
+			parentEnv = p.tags["env_name"]
+			if parentEnv != "-" {
+				break
+			}
+		}
+		for p := parent; p != nil; p = p.parent {
+			parentFlag = p.tags["flag_name"]
+			if parentFlag != "-" {
+				break
+			}
+		}
+
+		parentEnv += sp.cfg.envDelimiter
+		parentFlag += sp.cfg.FlagDelimiter
+	}
+
+	pfield := &parsedField{
+		name:     name,
+		namefull: parentName + name,
+		parent:   parent,
+		tags: map[string]string{
+			"usage":     field.Tag.Get("usage"),
+			"env_name":  env,
+			"env_full":  sp.cfg.EnvPrefix + parentEnv + env,
+			"flag_name": flag,
+			"flag_full": sp.cfg.FlagPrefix + parentFlag + flag,
+		},
+		isRequired: requiredTag == "true",
+	}
+
+	if !sp.cfg.SkipDefaults {
+		// TODO: must be typed?
+		pfield.defaultValue = field.Tag.Get("default")
+	}
+
+	if env == "-" {
+		delete(pfield.tags, "env_full")
+	}
+	if flag == "-" {
+		delete(pfield.tags, "flag_full")
+	}
+
+	if exactName, _, ok := strings.Cut(env, ",exact"); ok {
+		pfield.tags["env_full"] = exactName
+	}
+	if exactName, _, ok := strings.Cut(flag, ",exact"); ok {
+		pfield.tags["flag_full"] = exactName
+	}
+
+	if !sp.cfg.AllowDuplicates {
+		name := pfield.tags["env_full"]
+		if _, ok := sp.envNames[name]; ok && name != "" {
+			return nil, fmt.Errorf("field %q is duplicated", name)
+		}
+		sp.envNames[name] = struct{}{}
+	}
+
+	if !sp.cfg.SkipFlags {
+		flagName := pfield.tags["flag_full"]
+		if flagName != "" {
+			if _, ok := sp.flagNames[flagName]; ok && !sp.cfg.AllowDuplicates {
+				return nil, fmt.Errorf("duplicate flag %q", flagName)
+			}
+			sp.flagNames[flagName] = struct{}{}
+			// TODO: must be typed
+			sp.flagSet.String(flagName, field.Tag.Get("default"), field.Tag.Get("usage"))
+		}
+	}
+
+	if sp.cfg.DontGenerateTags {
+		newName = name
+	}
+	for _, dec := range sp.cfg.FileDecoders {
+		format := dec.Format()
+		v := field.Tag.Get(format)
+		if v == "" {
+			v = newName
+		}
+		pfield.tags[format] = v
+	}
+	return pfield, nil
+}
+
+func (sp *structParser) parseStruct(x interface{}) error {
+	value := reflect.ValueOf(x)
+	if value.Type().Kind() == reflect.Ptr {
+		value = value.Elem()
+	}
+
+	fields, err := sp.parseStructHelper(nil, value, map[string]interface{}{})
+	if err != nil {
+		return err
+	}
+	sp.fields = fields
+
+	// fmt.Printf("fields: %+v\n", fields)
+	return nil
+}
+
+func (sp *structParser) parseStructHelper(parent *parsedField, structValue reflect.Value, res map[string]interface{}) (map[string]interface{}, error) {
+	count := structValue.NumField()
+	structType := structValue.Type()
+
+	for i := 0; i < count; i++ {
+		field := structType.Field(i)
+		fieldValue := structValue.Field(i)
+		fieldType := fieldValue.Type()
+		if !fieldValue.CanSet() {
+			continue
+		}
+
+		defaultTagValue := field.Tag.Get("default")
+		pfield, err := sp.newParseField(parent, field)
+		if err != nil {
+			return nil, err
+		}
+
+		// do not set defaultValue for struct or pointer type without a default value
+		// if fieldType.Kind() == reflect.Struct ||
+		// 	(fieldType.Kind() == reflect.Pointer && defaultTagValue == "") {
+		// 	pfield.defaultValue = nil
+		// }
+
+		if fieldType.Kind() == reflect.Pointer {
+			fieldValue = fieldValue.Elem()
+			fieldValue = reflect.New(fieldType)
+			fieldType = fieldValue.Type()
+		}
+
+		value := fieldValue.Interface() // to have 'value' of type field
+
+		// if !sp.cfg.SkipDefaults {
+		// 	pv := fieldValue.Addr().Interface()
+		// 	if v, ok := pv.(encoding.TextUnmarshaler); ok {
+		// 		value = defaultTagValue
+		// 		err := v.UnmarshalText([]byte(fmt.Sprint(value)))
+		// 		if err != nil {
+		// 			return nil, err
+		// 		}
+		// 	}
+		// 	pfield.value =
+		// 	res[pfield.name] = pfield
+		// 	continue
+		// }
+
+		switch fieldType.Kind() {
+		// case reflect.Array:
+		// TODO: same as slice + check len?
+
+		case reflect.Interface:
+			// TODO: just assign?
+
+		case reflect.Struct:
+			pfield.hasChilds = true
+
+			param := map[string]interface{}{}
+			parent := pfield
+			if field.Anonymous {
+				pfield.hasChilds = false
+				param = res
+				parent = pfield.parent
+			}
+
+			values, err := sp.parseStructHelper(parent, fieldValue, param)
+			if err != nil {
+				return nil, err
+			}
+			// fmt.Printf("field: %+v got: %+v\n\n", pfield.name, values)
+
+			value = values
+
+		case reflect.Slice, reflect.Array:
+			if isPrimitive(field.Type.Elem()) {
+				// byte-slice case
+				if field.Type.Elem().Kind() == reflect.Uint8 {
+					value = []byte(defaultTagValue)
+				} else {
+					values := []interface{}{}
+					if defaultTagValue != "" && strings.Index(defaultTagValue, ",") == -1 {
+						return nil, fmt.Errorf("incorrect default tag value for slice/array: %v", defaultTagValue)
+					}
+					for _, val := range strings.Split(defaultTagValue, ",") {
+						values = append(values, val)
+					}
+					value = values
+				}
+			} else {
+				pfield.hasChilds = true
+				// TODO: if value is struct - parse
+				// value = parseSlice(fieldValue, map[string]interface{}{})
+			}
+
+			// if !sp.cfg.SkipDefaults {
+			// 	pfield.value = value
+			// }
+
+		case reflect.Map:
+			// if isPrimitive(field.Type.Elem()) {
+			values := map[string]interface{}{}
+			parts := strings.Split(defaultTagValue, ",")
+			if defaultTagValue != "" && strings.Index(defaultTagValue, ",") == -1 {
+				return nil, fmt.Errorf("incorrect default tag value for map: %v", defaultTagValue)
+			}
+
+			if len(parts) > 1 {
+				for _, entry := range parts {
+					// fmt.Printf("parts: %+v\n", parts)
+					entries := strings.SplitN(entry, ":", 2)
+					if len(entries) != 2 {
+						return nil, fmt.Errorf("want 2 parts got %d (%s)", len(entries), entries)
+					}
+					// TODO: convert entry[1] to a primitive?
+					values[entries[0]] = entries[1]
+				}
+			}
+			value = values
+			// } else {
+			// 	pfield.hasChilds = true
+			// }
+
+		default:
+			// TODO: do not set pointer
+			if fieldType.Kind() == reflect.Pointer && defaultTagValue == "" {
+				// skip
+				value = nil
+			} else {
+				// TODO: when WeaklyTypedInput will be false use decodePrimitive(...)
+				if !sp.cfg.SkipDefaults {
+					value = defaultTagValue
+					if fieldType == reflect.TypeOf(time.Second) {
+						val, err := time.ParseDuration(defaultTagValue)
+						if err != nil {
+							return nil, err
+						}
+						value = val
+					}
+				}
+
+			}
+		}
+
+		// we should not overwrite struct because there are childs
+		if sp.cfg.SkipDefaults && fieldType.Kind() != reflect.Struct {
+			pfield.value = fieldValue.Interface()
+		} else {
+			pfield.value = value
+		}
+
+		// fmt.Printf("def: %v %T '%+v'\n", fieldType.String(), value, value)
+		res[pfield.name] = pfield
+	}
+	return res, nil
+}
+
+var fieldType = reflect.TypeOf(&parsedField{})
+
+var hook = mapstructure.DecodeHookFuncType(func(from, to reflect.Type, data interface{}) (interface{}, error) {
+	if from != fieldType {
+		// fmt.Printf("hook: got %T (%+v) when %s\n", i, i, to.String())
+		return data, nil
+	}
+	field := data.(*parsedField)
+
+	ifaceTo := reflect.New(to).Interface()
+	if unmarshaller, ok := ifaceTo.(encoding.TextUnmarshaler); ok {
+		// TODO: only string can be here?
+		b := []byte(field.value.(string))
+		err := unmarshaller.UnmarshalText(b)
+		return unmarshaller, err
+	}
+	// fmt.Printf("hook: when %s do '%+v' // %+v\n\n", to.String(), field.value, field)
+	return field.value, nil
+})
+
+func (sp *structParser) apply(x interface{}) error {
+	dec, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
+		Result:           x,
+		DecodeHook:       hook,
+		WeaklyTypedInput: true, // TODO: temp fix?
+	})
+	if err != nil {
+		panic(fmt.Sprintf("aconfig: BUG with mapstructure.NewDecoder: %v", err))
+	}
+
+	if err := dec.Decode(sp.fields); err != nil {
+		return fmt.Errorf("decode: %w", err)
+	}
+	return nil
+}
+
+func (sp *structParser) applyLevel(tag string, values map[string]interface{}) error {
+	if err := sp.applyLevelHelper2(sp.fields, tag, values); err != nil {
+		return err
+	}
+
+	if !sp.cfg.AllowUnknownFields {
+		for env, value := range values {
+			return fmt.Errorf("unknown field in file %q: %s=%v (see AllowUnknownFields config param)", "file", env, value)
+		}
+	}
+	return nil
+}
+
+func (sp *structParser) applyLevelHelper2(fields map[string]interface{}, tag string, values map[string]interface{}) error {
+	for _, field := range fields {
+		pfield, ok := field.(*parsedField)
+		if !ok {
+			fmt.Printf("wat in level %T (%+v)\n", field, field)
+			continue
+		}
+		tagValue, ok := pfield.tags[tag]
+		if !ok {
+			continue
+		}
+		value, ok := values[tagValue]
+		if !ok {
+			continue
+		}
+
+		switch value := value.(type) {
+		case map[string]interface{}:
+			if pfield.hasChilds {
+				pfieldValue, ok := pfield.value.(map[string]interface{})
+				if !ok {
+					fmt.Printf("ouch %T (%+v)\n", pfield.value, pfield.value)
+					continue
+				}
+				err := sp.applyLevelHelper2(pfieldValue, tag, value)
+				if err != nil {
+					return err
+				}
+			} else {
+				pfield.value = value
+			}
+		default:
+			pfield.value = value
+		}
+
+		delete(values, tagValue)
+	}
+	return nil
+}
+
+func (sp *structParser) applyLevelHelper(fields map[string]interface{}, tag string, values map[string]interface{}) error {
+	for _, v := range fields {
+		field, ok := v.(*parsedField)
+		if !ok {
+			// fmt.Printf("got type %T (%v)\n", v, v)
+			continue
+		}
+
+		want := field.tags[tag]
+		value, ok := values[want]
+		if !ok {
+			continue
+		}
+		vval, ok := value.(map[string]interface{})
+
+		// TODO: can be only for leaf nodes?
+		if !ok {
+			// fmt.Printf("got val %T (%v)\n", val, val)
+			field.value = value
+			continue
+		}
+
+		// fmt.Printf("got map: %+v %T\n", vval, vval)
+
+		// no struct in childs - simple apply, mapstructure will take care
+		if field.childs == nil {
+			// TODO: reencode values?
+			field.value = vval
+		} else {
+			if err := sp.applyLevelHelper(field.childs, tag, vval); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (sp *structParser) applyFlat(tag string, values map[string]interface{}) error {
+	allowUnknown := true
+	prefix := ""
+
+	switch tag {
+	case "env":
+		allowUnknown, prefix = sp.cfg.AllowUnknownEnvs, sp.cfg.EnvPrefix
+	case "flag":
+		allowUnknown, prefix = sp.cfg.AllowUnknownFlags, sp.cfg.FlagPrefix
+	}
+
+	dupls := map[string]struct{}{}
+
+	if err := sp.applyFlatHelper(sp.fields, tag, values); err != nil {
+		return err
+	}
+
+	if allowUnknown || prefix == "" {
+		return nil
+	}
+
+	for name := range dupls {
+		delete(values, name)
+	}
+	for key, value := range values {
+		if strings.HasPrefix(key, prefix) {
+			return fmt.Errorf("unknown %s %s=%v (see AllowUnknownXXX config param)", tag, key, value)
+		}
+	}
+	return nil
+}
+
+func (sp *structParser) applyFlatHelper(fields map[string]interface{}, tag string, values map[string]interface{}) error {
+	for _, field := range fields {
+		pfield, ok := field.(*parsedField)
+		if !ok {
+			fmt.Printf("wat in flat %T (%+v)\n", field, field)
+			continue
+		}
+
+		tagValue, ok := pfield.tags[tag+"_full"]
+		if !ok {
+			continue
+		}
+		value, ok := values[tagValue]
+		if !ok {
+			if !pfield.hasChilds {
+				continue
+			}
+			if err := sp.applyFlatHelper(pfield.value.(map[string]interface{}), tag, values); err != nil {
+				return err
+			}
+			continue
+		}
+
+		pfield.value = value
+		if !sp.cfg.AllowDuplicates {
+			delete(values, tagValue)
+		}
+	}
+	return nil
+}
+
+func isPrimitive(v reflect.Type) bool {
+	return v.Kind() < reflect.Array || v.Kind() == reflect.String
+}

--- a/parser.go
+++ b/parser.go
@@ -12,9 +12,8 @@ import (
 )
 
 type structParser struct {
-	dst       interface{}
 	cfg       Config
-	fields    map[string]interface{}
+	fields    map[string]any
 	flagSet   *flag.FlagSet
 	envNames  map[string]struct{}
 	flagNames map[string]struct{}
@@ -32,14 +31,13 @@ func newStructParser(cfg Config) *structParser {
 type parsedField struct {
 	name         string
 	namefull     string
-	value        interface{}
-	defaultValue interface{}
+	value        any
+	defaultValue any
 	parent       *parsedField
-	childs       map[string]interface{}
+	childs       map[string]any
 	tags         map[string]string
 	hasChilds    bool
 	isRequired   bool
-	isSet        bool
 }
 
 func (pf *parsedField) String() string {
@@ -160,13 +158,13 @@ func (sp *structParser) newParseField(parent *parsedField, field reflect.StructF
 	return pfield, nil
 }
 
-func (sp *structParser) parseStruct(x interface{}) error {
+func (sp *structParser) parseStruct(x any) error {
 	value := reflect.ValueOf(x)
 	if value.Type().Kind() == reflect.Ptr {
 		value = value.Elem()
 	}
 
-	fields, err := sp.parseStructHelper(nil, value, map[string]interface{}{})
+	fields, err := sp.parseStructHelper(nil, value, map[string]any{})
 	if err != nil {
 		return err
 	}
@@ -176,7 +174,7 @@ func (sp *structParser) parseStruct(x interface{}) error {
 	return nil
 }
 
-func (sp *structParser) parseStructHelper(parent *parsedField, structValue reflect.Value, res map[string]interface{}) (map[string]interface{}, error) {
+func (sp *structParser) parseStructHelper(parent *parsedField, structValue reflect.Value, res map[string]any) (map[string]any, error) {
 	count := structValue.NumField()
 	structType := structValue.Type()
 
@@ -232,7 +230,7 @@ func (sp *structParser) parseStructHelper(parent *parsedField, structValue refle
 		case reflect.Struct:
 			pfield.hasChilds = true
 
-			param := map[string]interface{}{}
+			param := map[string]any{}
 			parent := pfield
 			if field.Anonymous {
 				pfield.hasChilds = false
@@ -254,8 +252,8 @@ func (sp *structParser) parseStructHelper(parent *parsedField, structValue refle
 				if field.Type.Elem().Kind() == reflect.Uint8 {
 					value = []byte(defaultTagValue)
 				} else {
-					values := []interface{}{}
-					if defaultTagValue != "" && strings.Index(defaultTagValue, ",") == -1 {
+					values := []any{}
+					if defaultTagValue != "" && !strings.Contains(defaultTagValue, ",") {
 						return nil, fmt.Errorf("incorrect default tag value for slice/array: %v", defaultTagValue)
 					}
 					for _, val := range strings.Split(defaultTagValue, ",") {
@@ -266,7 +264,7 @@ func (sp *structParser) parseStructHelper(parent *parsedField, structValue refle
 			} else {
 				pfield.hasChilds = true
 				// TODO: if value is struct - parse
-				// value = parseSlice(fieldValue, map[string]interface{}{})
+				// value = parseSlice(fieldValue, map[string]any{})
 			}
 
 			// if !sp.cfg.SkipDefaults {
@@ -275,9 +273,9 @@ func (sp *structParser) parseStructHelper(parent *parsedField, structValue refle
 
 		case reflect.Map:
 			// if isPrimitive(field.Type.Elem()) {
-			values := map[string]interface{}{}
+			values := map[string]any{}
 			parts := strings.Split(defaultTagValue, ",")
-			if defaultTagValue != "" && strings.Index(defaultTagValue, ",") == -1 {
+			if defaultTagValue != "" && !strings.Contains(defaultTagValue, ",") {
 				return nil, fmt.Errorf("incorrect default tag value for map: %v", defaultTagValue)
 			}
 
@@ -314,7 +312,6 @@ func (sp *structParser) parseStructHelper(parent *parsedField, structValue refle
 						value = val
 					}
 				}
-
 			}
 		}
 
@@ -333,7 +330,7 @@ func (sp *structParser) parseStructHelper(parent *parsedField, structValue refle
 
 var fieldType = reflect.TypeOf(&parsedField{})
 
-var hook = mapstructure.DecodeHookFuncType(func(from, to reflect.Type, data interface{}) (interface{}, error) {
+var hook = mapstructure.DecodeHookFuncType(func(from, to reflect.Type, data any) (any, error) {
 	if from != fieldType {
 		// fmt.Printf("hook: got %T (%+v) when %s\n", i, i, to.String())
 		return data, nil
@@ -351,7 +348,7 @@ var hook = mapstructure.DecodeHookFuncType(func(from, to reflect.Type, data inte
 	return field.value, nil
 })
 
-func (sp *structParser) apply(x interface{}) error {
+func (sp *structParser) apply(x any) error {
 	dec, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
 		Result:           x,
 		DecodeHook:       hook,
@@ -367,7 +364,7 @@ func (sp *structParser) apply(x interface{}) error {
 	return nil
 }
 
-func (sp *structParser) applyLevel(tag string, values map[string]interface{}) error {
+func (sp *structParser) applyLevel(tag string, values map[string]any) error {
 	if err := sp.applyLevelHelper2(sp.fields, tag, values); err != nil {
 		return err
 	}
@@ -380,7 +377,7 @@ func (sp *structParser) applyLevel(tag string, values map[string]interface{}) er
 	return nil
 }
 
-func (sp *structParser) applyLevelHelper2(fields map[string]interface{}, tag string, values map[string]interface{}) error {
+func (sp *structParser) applyLevelHelper2(fields map[string]any, tag string, values map[string]any) error {
 	for _, field := range fields {
 		pfield, ok := field.(*parsedField)
 		if !ok {
@@ -397,9 +394,9 @@ func (sp *structParser) applyLevelHelper2(fields map[string]interface{}, tag str
 		}
 
 		switch value := value.(type) {
-		case map[string]interface{}:
+		case map[string]any:
 			if pfield.hasChilds {
-				pfieldValue, ok := pfield.value.(map[string]interface{})
+				pfieldValue, ok := pfield.value.(map[string]any)
 				if !ok {
 					fmt.Printf("ouch %T (%+v)\n", pfield.value, pfield.value)
 					continue
@@ -420,7 +417,7 @@ func (sp *structParser) applyLevelHelper2(fields map[string]interface{}, tag str
 	return nil
 }
 
-func (sp *structParser) applyLevelHelper(fields map[string]interface{}, tag string, values map[string]interface{}) error {
+func (sp *structParser) applyLevelHelper(fields map[string]any, tag string, values map[string]any) error {
 	for _, v := range fields {
 		field, ok := v.(*parsedField)
 		if !ok {
@@ -433,7 +430,7 @@ func (sp *structParser) applyLevelHelper(fields map[string]interface{}, tag stri
 		if !ok {
 			continue
 		}
-		vval, ok := value.(map[string]interface{})
+		vval, ok := value.(map[string]any)
 
 		// TODO: can be only for leaf nodes?
 		if !ok {
@@ -457,7 +454,7 @@ func (sp *structParser) applyLevelHelper(fields map[string]interface{}, tag stri
 	return nil
 }
 
-func (sp *structParser) applyFlat(tag string, values map[string]interface{}) error {
+func (sp *structParser) applyFlat(tag string, values map[string]any) error {
 	allowUnknown := true
 	prefix := ""
 
@@ -489,7 +486,7 @@ func (sp *structParser) applyFlat(tag string, values map[string]interface{}) err
 	return nil
 }
 
-func (sp *structParser) applyFlatHelper(fields map[string]interface{}, tag string, values map[string]interface{}) error {
+func (sp *structParser) applyFlatHelper(fields map[string]any, tag string, values map[string]any) error {
 	for _, field := range fields {
 		pfield, ok := field.(*parsedField)
 		if !ok {
@@ -506,7 +503,7 @@ func (sp *structParser) applyFlatHelper(fields map[string]interface{}, tag strin
 			if !pfield.hasChilds {
 				continue
 			}
-			if err := sp.applyFlatHelper(pfield.value.(map[string]interface{}), tag, values); err != nil {
+			if err := sp.applyFlatHelper(pfield.value.(map[string]any), tag, values); err != nil {
 				return err
 			}
 			continue

--- a/reflection.go
+++ b/reflection.go
@@ -404,7 +404,3 @@ func mii(m interface{}) map[string]interface{} {
 		panic(fmt.Sprintf("%T %v", m, m))
 	}
 }
-
-func isPrimitive(v reflect.Type) bool {
-	return v.Kind() < reflect.Array || v.Kind() == reflect.String
-}

